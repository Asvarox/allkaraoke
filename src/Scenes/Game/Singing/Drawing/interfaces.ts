export default interface Particle {
    tick: (ctx: CanvasRenderingContext2D, canvas: HTMLCanvasElement) => void;
    finished: boolean;
}
