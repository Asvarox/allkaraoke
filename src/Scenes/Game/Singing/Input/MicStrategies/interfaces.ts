export interface FrequencyDetectionStrategy {
    getSampleSize(): number;

    init(context: AudioContext, processor: ScriptProcessorNode): Promise<void>;

    getFrequency(data: Float32Array): Promise<number>;
}
