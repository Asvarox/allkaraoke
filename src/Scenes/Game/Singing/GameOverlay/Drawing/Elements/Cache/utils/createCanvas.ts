// In tests this function is mocked so it creates node-canvas object
export default function createCanvas(w: number, h: number) {
    const canvas = document.createElement('canvas');
    canvas.width = w;
    canvas.height = h;

    return canvas;
}
