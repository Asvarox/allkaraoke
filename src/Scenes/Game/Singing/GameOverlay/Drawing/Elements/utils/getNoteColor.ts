import { PlayerNote } from 'interfaces';
import styles from '../../styles';

export default function getNoteColor(
    ctx: CanvasRenderingContext2D,
    playerNumber: number,
    isHit: boolean,
    playerNote: PlayerNote,
) {
    if (playerNote.isPerfect && playerNote.note.type === 'star') {
        return styles.colors.players[playerNumber].starPerfect;
    } else if (playerNote.isPerfect) {
        return styles.colors.players[playerNumber].perfect;
    } else if (playerNote.note.type === 'star' && isHit) {
        return styles.colors.players[playerNumber].star;
    } else if (isHit) {
        return styles.colors.players[playerNumber].hit;
    } else {
        return styles.colors.players[playerNumber].miss;
    }
}
