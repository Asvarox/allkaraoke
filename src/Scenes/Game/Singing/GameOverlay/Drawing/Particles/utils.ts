const spreadValue = (initial: number, spread: number) => initial - spread + Math.random() * spread * 2;

export default spreadValue;

export const randomSign = (weight = 0) => Math.sign(weight + Math.random() - 0.5);
