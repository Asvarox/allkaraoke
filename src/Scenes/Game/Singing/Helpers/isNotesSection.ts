import { NotesSection, Section } from '../../../../interfaces';

export default function isNotesSection(section: Section | undefined | null): section is NotesSection {
    return section?.type === 'notes' && !!section.notes.length;
}
