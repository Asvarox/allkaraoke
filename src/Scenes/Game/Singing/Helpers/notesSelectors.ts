import { Section } from "../../../../interfaces";
import isNotesSection from "./isNotesSection";

export const getFirstNoteStartFromSections = (sections: Section[]) => {
    const firstNoteSection = sections.find(isNotesSection);

    return firstNoteSection ? firstNoteSection.notes[0].start : Infinity;
}
export const getLastNoteEndFromSections = (sections: Section[]) => {
    const notesSections = sections.filter(isNotesSection);
    const lastSection = notesSections[notesSections.length - 1];
    const lastNote = lastSection.notes[lastSection.notes.length - 1];

    return lastNote.start + lastNote.length;
}