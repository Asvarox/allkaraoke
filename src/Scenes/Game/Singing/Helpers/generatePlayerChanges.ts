import { NotesSection, Section, Song } from 'interfaces';
import { cloneDeep, range } from 'lodash-es';
import getSongBeatLength from '../GameState/Helpers/getSongBeatLength';
import isNotesSection from './isNotesSection';
import { getFirstNoteStartFromSections, getLastNoteEndFromSections } from './notesSelectors';

const MAX_SECTION_PAUSE_MS = 250;

const getSingableBeats = (sections: Section[], start: number = 0, end: number = Infinity) =>
    sections
        .filter(isNotesSection)
        .filter((section) => getFirstNoteStartFromSections([section]) >= start)
        .filter((section) => getLastNoteEndFromSections([section]) <= end)
        .reduce((current, section) => current + section.notes.reduce((length, note) => length + note.length, 0), 0);

// Join sections into segments - group of sections that have shorter than MAX_SECTION_PAUSE_MS pause between
export const segmentSections = (sections: Section[], beatLength: number): NotesSection[][] => {
    const [firstSection, ...noteSections] = sections.filter(isNotesSection);

    const segments: NotesSection[][] = [[firstSection]];

    noteSections.forEach((section) => {
        const lastSegment = segments.at(-1)!;

        const lastSectionEnd = getLastNoteEndFromSections(lastSegment);
        const firstSectionNoteStart = getFirstNoteStartFromSections([section]);
        const difference = firstSectionNoteStart - lastSectionEnd;

        if (difference * beatLength > MAX_SECTION_PAUSE_MS) {
            segments.push([section]);
        } else {
            lastSegment.push(section);
        }
    });

    return segments;
};

const sum = (acc: number, elem: number) => acc + elem;

const calculateAmplitude = (groups: number[][]) => {
    const sums = groups.map((group) => group.reduce(sum, 0));
    return Math.max(...sums) - Math.min(...sums);
};

const shiftSolution = (groups: number[][], at: number): number[][] => {
    const result = cloneDeep(groups);
    const takeFromLeft = (pos: number) => {
        if (result[pos - 1].length < 2) return;
        const left = result[pos - 1].pop()!;
        result[pos].push(left);
    };
    const takeFromRight = (pos: number) => {
        if (result[pos + 1]?.length < 2) return;
        const [first, ...rest] = result[pos + 1];
        result[pos].push(first);
        result[pos + 1] = rest;
    };

    at % 2 === 1 ? takeFromLeft(Math.ceil(at / 2)) : takeFromRight(Math.ceil(at / 2));

    return result;
};

const arrSum = (arr: number[]) => arr.reduce(sum, 0);

export default function generatePlayerChanges(song: Song): number[][] {
    const CHANGES_COUNT = 9;
    const beatLength = getSongBeatLength(song);

    return song.tracks.map((track) => {
        const segments = segmentSections(track.sections, beatLength);

        // array of groups of singable beats count -- each element is the total singable beats in a segment
        // The goal is to for the groups to have as similar total count of singable beats as possible
        let result: number[][] = [];

        // todo replace this hacky algorithm with https://en.wikipedia.org/wiki/Multiway_number_partitioning
        // STEP1:   Build initial groups by calculating "ideal" group length (aka amount of singable beats)
        //          And try to have them as close to the ideal as possible (by either overflowing or "underflowing")
        let currentBuffer = [getSingableBeats(segments[0])]; // A candidate for the next group
        for (let i = 0; i < segments.length - 2; i++) {
            // get the last segment already put in the result
            const lastSegment = segments[result.flat().length - 1] ?? [];
            // How many beats are left to sing
            const singableBeats = getSingableBeats(track.sections, getLastNoteEndFromSections(lastSegment) ?? 0);
            const idealSize = singableBeats / (1 + CHANGES_COUNT - result.length);

            const segmentSingableBeatsCount = getSingableBeats(segments[i + 1]);

            const bufferSum = arrSum(currentBuffer);

            if (bufferSum + segmentSingableBeatsCount < idealSize) {
                // If adding the current segment to the candidate doesn't move us past the ideal size, just add it
                currentBuffer = [...currentBuffer, segmentSingableBeatsCount];
            } else if (idealSize - bufferSum < bufferSum + segmentSingableBeatsCount - idealSize) {
                // If the candidate is closer to ideal size, add it as a new group
                result.push(currentBuffer);
                currentBuffer = [segmentSingableBeatsCount];
            } else {
                // If the candidate with currently inspected segment is closer to ideal size, add it as a new group
                result.push([...currentBuffer, segmentSingableBeatsCount]);
                currentBuffer = [];
            }
        }

        // STEP2:   Try to make groups more even by trying to move segments between groups and checking if it
        //          resulted in smaller differences between longest and shortest group (aka amplitude)
        for (let i = 0; i < 100; i++) {
            const currentAmplitude = calculateAmplitude(result);
            // Create alternative results (by shifting one segment between groups in either way)
            // Note this doesn't guarantee creating best results as moving multiple segments at the same time
            // isn't accounted for
            const candidates = range((result.length - 2) * 2 + 2).map((at) => shiftSolution(result, at));
            const candidatesAmplitudes = candidates.map(calculateAmplitude);
            const minimalAmplitude = Math.min(...candidatesAmplitudes);
            if (currentAmplitude <= minimalAmplitude) break; // If no results are better, give up
            const minimalAmplitudeIndex = candidatesAmplitudes.indexOf(minimalAmplitude);
            result = candidates[minimalAmplitudeIndex];
        }

        return result
            .map((group) => group.length)
            .reduce<number[]>((acc, length) => [...acc, (acc.at(-1) ?? 0) + length], [])
            .map((index) => getLastNoteEndFromSections(segments[index - 1]));
    });
}
