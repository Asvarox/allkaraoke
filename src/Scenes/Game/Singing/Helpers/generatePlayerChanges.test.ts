import { Section, Song } from 'interfaces';
import isNotesSection from 'Songs/utils/isNotesSection';
import { getFirstNoteStartFromSections, getLastNoteEndFromSections } from 'Songs/utils/notesSelectors';
import { generateSection, generateSong } from 'utils/testUtils';
import generatePlayerChanges from './generatePlayerChanges';

const getSingableBeats = (sections: Section[], start: number, end: number) =>
    sections
        .filter(isNotesSection)
        .filter((section) => getFirstNoteStartFromSections([section]) >= start)
        .filter((section) => getLastNoteEndFromSections([section]) <= end)
        .reduce((current, section) => current + section.notes.reduce((length, note) => length + note.length, 0), 0);

describe('generatePlayerChanges', () => {
    it('should generate 10 parts', () => {
        const song = generateSong(
            [
                [
                    generateSection(4 * 0, 1, 1),
                    generateSection(4 * 1, 1, 1),
                    generateSection(4 * 2, 1, 1),
                    generateSection(4 * 3, 1, 1),
                    generateSection(4 * 4, 1, 1),
                    generateSection(4 * 5, 1, 1),
                    generateSection(4 * 6, 1, 1),
                    generateSection(4 * 7, 1, 1),
                    generateSection(4 * 8, 1, 1),
                    generateSection(4 * 9, 1, 1),
                    generateSection(4 * 10, 1, 1),
                    generateSection(4 * 11, 1, 1),
                    generateSection(4 * 12, 1, 1),
                    generateSection(4 * 13, 1, 1),
                    generateSection(4 * 14, 1, 1),
                    generateSection(4 * 15, 1, 1),
                    generateSection(4 * 16, 1, 1),
                    generateSection(4 * 17, 1, 1),
                    generateSection(4 * 18, 1, 1),
                    generateSection(4 * 19, 1, 1),
                ],
            ],
            { bpm: 0.1 },
        );

        expect(generatePlayerChanges(song)).toEqual([
            [4 + 1, 4 * 3 + 1, 4 * 5 + 1, 4 * 7 + 1, 4 * 9 + 1, 4 * 11 + 1, 4 * 13 + 1, 4 * 15 + 1, 4 * 17 + 1],
        ]);
    });

    it.each([
        [require('../../../../../public/songs/Big Cyc-Swiat Wedlug Kiepskich.json')],
        [require('../../../../../public/songs/Cailin Russo & Chrissy Costanza-Phoenix.json')],
        [require('../../../../../public/songs/T.Love-King.json')],
        [require('../../../../../public/songs/Ed Sheeran-Galway Girl.json')],
        [require('../../../../../public/songs/Muse-Madness.json')],
        [require('../../../../../public/songs/Jeden Osiem L-Jak zapomniec.json')],
    ])('should handle real life examples better', (song: Song) => {
        expect(generatePlayerChanges(song)).toMatchSnapshot();
    });
});
