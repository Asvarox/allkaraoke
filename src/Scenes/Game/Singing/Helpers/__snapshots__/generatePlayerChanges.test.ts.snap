// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generatePlayerChanges should handle real life examples better 1`] = `
Array [
  Array [
    73,
    136,
    231,
    323,
    551,
    648,
    743,
    835,
    935,
  ],
]
`;

exports[`generatePlayerChanges should handle real life examples better 2`] = `
Array [
  Array [
    291,
    566,
    998,
    1290,
    1655,
    1974,
    2406,
    2723,
    3687,
  ],
]
`;

exports[`generatePlayerChanges should handle real life examples better 3`] = `
Array [
  Array [
    184,
    378,
    572,
    767,
    1210,
    1401,
    1594,
    1784,
    2842,
  ],
]
`;

exports[`generatePlayerChanges should handle real life examples better 4`] = `
Array [
  Array [
    667,
    916,
    1222,
    1547,
    1938,
    2245,
    2703,
    3086,
    3399,
  ],
]
`;

exports[`generatePlayerChanges should handle real life examples better 5`] = `
Array [
  Array [
    586,
    1182,
    1810,
    2458,
    3094,
    4374,
    5052,
    5572,
    6078,
  ],
]
`;

exports[`generatePlayerChanges should handle real life examples better 6`] = `
Array [
  Array [
    457,
    861,
    1285,
    1687,
    2117,
    2565,
    3017,
    3489,
    3991,
  ],
]
`;
